src/text_atlas.rs

```diff
@@ -68,20 +76,121 @@ impl InnerAtlas {
 
         loop {
             let allocation = self.packer.allocate(size);
+
             if allocation.is_some() {
                 return allocation;
             }
 
             // Try to free least recently used allocation
-            let (_, value) = self.glyph_cache.pop_lru()?;
-            self.packer
-                .deallocate(value.atlas_id.expect("cache corrupt"));
+            let (mut key, mut value) = self.glyph_cache.peek_lru()?;
+
+            while value.atlas_id.is_none() {
+                let _ = self.glyph_cache.pop_lru();
+
+                (key, value) = self.glyph_cache.peek_lru()?;
+            }
+
+            if self.glyphs_in_use.contains(&key) {
+                return None;
+            }
+
+            let (_, value) = self.glyph_cache.pop_lru().unwrap();
+            self.packer.deallocate(value.atlas_id.unwrap());
         }
     }
 
     pub fn num_channels(&self) -> usize {
         self.kind.num_channels()
     }
+
+    pub(crate) fn promote(&mut self, glyph: CacheKey) {
+        self.glyph_cache.promote(&glyph);
+        self.glyphs_in_use.insert(glyph);
+    }
+
+    pub(crate) fn put(&mut self, glyph: CacheKey, details: GlyphDetails) {
+        self.glyph_cache.put(glyph, details);
+        self.glyphs_in_use.insert(glyph);
+    }
+
+    pub(crate) fn grow(
+        &mut self,
+        device: &wgpu::Device,
+        queue: &wgpu::Queue,
+        font_system: &mut FontSystem,
+        cache: &mut SwashCache,
+    ) -> bool {
+        if self.size >= self.max_texture_dimension_2d {
+            return false;
+        }
+
+        // TODO: Better resizing logic (?)
+        let new_size = (self.size + Self::INITIAL_SIZE).min(self.max_texture_dimension_2d);
```

Comment:
A growth factor might be nice to avoid a lot of small allocations if the initial size is small, but we could add that later. e.g., maybe defaulting to 1.5x or 2x

Reply:
Yeah. Let's add that later. The new `grow` operation in `etagere` works well with multiples of the initial size.